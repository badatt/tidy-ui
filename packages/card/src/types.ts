import { HTMLAttributes } from 'react';
import { TColor } from '@tidy-ui/types';

export interface ICardProps extends HTMLAttributes<HTMLDivElement> {
  accent?: TColor;
  rounded?: boolean;
  href?: string;
}

export interface ICardFooterProps extends HTMLAttributes<HTMLDivElement> {
  separated?: boolean;
}

export interface ICardHeaderProps extends HTMLAttributes<HTMLDivElement> {
  separated?: boolean;
  href?: string;
}
