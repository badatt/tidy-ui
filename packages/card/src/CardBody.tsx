import { styled } from '@tidy-ui/theme';

/**
 * Card body component, powered by various styles (with theming) and settings
 *
 */
const CardBody = styled.div`
  padding: 0.5rem 1rem;
  display: flex;
`;

export { CardBody };
