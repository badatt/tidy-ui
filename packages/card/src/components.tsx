import { LaunchIcon } from '@tidy-ui/commons';
import { css, styled } from '@tidy-ui/theme';
import { ICardFooterProps, ICardHeaderProps, ICardProps } from './types';

/**
 * Internal CardHeader root component
 *
 * @internal
 */
const CardHeaderRoot = styled.header<ICardHeaderProps>`
  ${({ theme: { typography, palette }, isDivided }) => css`
    position: relative;
    font-weight: ${typography.fontWeightBold};
    padding-top: 1rem;
    padding-right: 1rem;
    padding-bottom: 0.5rem;
    padding-left: 1rem;
    width: 100%;
    ${isDivided &&
    css`
      border-bottom: 1px solid ${palette.divider};
    `}
  `}
`;

/**
 * Internal styled anchor tag
 *
 * @internal
 */
const Link = styled.a`
  display: none;
  position: absolute;
  height: 2.5rem;
  width: 2.5rem;
  top: 0;
  right: 0;
`;

/**
 * Internal styled LaunchIcon
 *
 * @internal
 */
const Icon = styled(LaunchIcon)`
  ${({ theme: { palette, isDark } }) => css`
    height: 1.25rem;
    width: 1.25rem;
    top: 0.5rem;
    right: 0.5rem;
    position: absolute;
    color: ${isDark ? palette['neutral'][500] : palette['neutral'][400]};
  `}
`;

/**
 * Internal Card root component
 *
 * @internal
 */
const CardRoot = styled.section<ICardProps>`
  ${({ theme: { palette, layout }, isSharp, withAccent, withHref }) => css`
    display: flex;
    flex-direction: column;
    background-color: ${palette.background.card};
    box-shadow: ${layout.shadow};
    color: ${palette.text.primary};
    margin-bottom: 1.875rem;
    ${!isSharp &&
    css`
      border-radius: ${layout.radius};
    `}
    ${withAccent &&
    css`
      border-top: 2px solid ${palette[withAccent][600]};
    `}
    ${withHref &&
    css`
      cursor: pointer;
    `}
  `}
  &:hover ${Link} {
    display: block;
  }
`;

/**
 * Internal CardFooter root component
 *
 * @internal
 */
const CardFooterRoot = styled.footer<ICardFooterProps>`
  display: flex;
  padding-top: 0.5rem;
  padding-right: 1rem;
  padding-bottom: 1rem;
  padding-left: 1rem;
  ${({ theme: { palette }, isDivided }) => css`
    ${isDivided &&
    css`
      border-top: 1px solid ${palette.divider};
    `}
  `}
`;

/**
 * Internal CardBody root component
 *
 * @internal
 */
const CardBodyRoot = styled.div`
  padding: 0.5rem 1rem;
  display: flex;
`;

export { CardBodyRoot, CardFooterRoot, CardHeaderRoot, CardRoot, Icon, Link };
