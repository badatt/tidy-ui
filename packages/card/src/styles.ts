import { createFontStyle, css } from '@tidy-ui/commons';
import { ICardProps } from './types';

const cardStyles = css<ICardProps>`
  ${({ theme: { palette, layout }, isSharp, accent, accentPosition, href, margin, height, width }) => css`
    ${createFontStyle()}
    background-color: ${palette.background.card};
    box-shadow: ${layout.shadow};
    color: ${palette.text.primary};
    margin: ${margin};
    height: ${height};
    width: ${width};
    ${!isSharp &&
    css`
      border-radius: ${layout.radius};
    `}
    ${accent && `border-${accentPosition}: 2px solid ${palette[accent][600]};`}
    ${href &&
    css`
      cursor: pointer;
    `}
  `}
`;

export { cardStyles };
