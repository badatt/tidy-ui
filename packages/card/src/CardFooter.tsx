import React from 'react';
import { CardFooterRoot } from './components';
import { ICardFooterProps } from './types';

/**
 * CardFooter component, powered by various styles (with theming) and settings
 *
 */
const CardFooter = React.forwardRef<HTMLDivElement, ICardFooterProps>((props, ref) => {
  const { children, ...rest } = props;
  return (
    <CardFooterRoot role="presentation" ref={ref} {...rest}>
      {children}
    </CardFooterRoot>
  );
});

CardFooter.defaultProps = {
  divided: false,
};

CardFooter.displayName = 'CardFooter';

export { CardFooter };
