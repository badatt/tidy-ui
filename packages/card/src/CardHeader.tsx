import React, { forwardRef } from 'react';
import { CardHeaderRoot, Icon, Link } from './components';
import { ICardHeaderProps } from './types';

/**
 * CardHeader component, powered by various styles (with theming) and settings
 *
 */
const CardHeader = forwardRef<HTMLDivElement, ICardHeaderProps>((props, ref) => {
  const { children, withHref, ...rest } = props;
  return (
    <CardHeaderRoot ref={ref} {...rest}>
      {children}
      {withHref && (
        <Link href={withHref} target="_blank">
          <Icon {...rest} />
        </Link>
      )}
    </CardHeaderRoot>
  );
});

CardHeader.defaultProps = {
  isDivided: false,
};

export { CardHeader };
