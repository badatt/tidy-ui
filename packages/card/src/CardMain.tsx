import React from 'react';
import { CardMainRoot } from './components';
import { ICardProps } from './types';

/**
 * Card component powered by various styles (with theming) and settings. Can be used just
 * as is with a nice elevation effect that looks like a card on screen
 *
 */
const CardMain = React.forwardRef<HTMLDivElement, ICardProps>((props, ref) => {
  const { children, href, ...rest } = props;
  return (
    <CardMainRoot role="article" ref={ref} href={href} {...rest}>
      {href ? <a href={href}>{children}</a> : <>{children}</>}
    </CardMainRoot>
  );
});

CardMain.defaultProps = {
  accentPosition: 'top',
  sharp: false,
};

CardMain.displayName = 'CardMain';

export { CardMain };
