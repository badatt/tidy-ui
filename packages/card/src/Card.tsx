import React from 'react';
import { CardBody } from './CardBody';
import { CardFooter } from './CardFooter';
import { CardHeader } from './CardHeader';
import { CardRoot } from './components';
import { ICardProps } from './types';

/** @internal */
interface CardComponent extends React.ForwardRefExoticComponent<ICardProps & React.RefAttributes<HTMLDivElement>> {
  /** @internal */
  Body: typeof CardBody;
  /** @internal */
  Footer: typeof CardFooter;
  /** @internal */
  Header: typeof CardHeader;
}

/**
 * Card component powered by various styles (with theming) and settings. Can be used just
 * as is with a nice elevation effect that looks like a card on screen
 *
 */
const Card = React.forwardRef<HTMLDivElement, ICardProps>((props, ref) => {
  const { children, ele, href, ...rest } = props;
  const nodes = ele ? React.cloneElement(ele, {}, children) : children;
  return (
    <CardRoot role="article" ref={ref} href={href} {...rest}>
      {href ? <a href={href}>{nodes}</a> : <>{nodes}</>}
    </CardRoot>
  );
}) as CardComponent;

Card.defaultProps = {
  accentPosition: 'top',
  isSharp: false,
};

Card.displayName = 'Card';

Card.Body = CardBody;
Card.Header = CardHeader;
Card.Footer = CardFooter;

export { Card };
