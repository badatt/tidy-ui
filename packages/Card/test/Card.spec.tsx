import React from "react";
import renderer from "react-test-renderer";
import "jest-styled-components";
import { Card } from "../src";

describe("Card", () => {
  test("Renders correctly", () => {
    const tree = renderer.create(<Card>This is a simple card</Card>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
