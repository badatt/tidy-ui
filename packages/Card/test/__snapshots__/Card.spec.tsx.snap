// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Renders correctly 1`] = `
<div>
  This is a simple card
</div>
`;
